### DISCUSSION TITLE
SQLite GRDB topics

### DISCUSSION THREAD

**Discussion by morstin** [pointfreeco/episode-code-samples#154] [Ideas]
@mbrandonw @stephencelis : Dear Brandon, Dear Stephen: thrilled that you are picking up SQLite / GRDB in your most recent episode. If you find time in this series, I would very interested in your take on:

- How to conform the table structs as Player, Team to Identifiable?
   <https://github.com/groue/GRDB.swift/issues/1435>
   When using those structs in SwiftUI Lists / IdentifiedArrays this would be very helpful.
   
- How to keep database consistency when deleting?
   <https://github.com/groue/GRDB.swift/discussions/1609>
   <https://forums.swift.org/t/grdb-persistence-callback-diddelete-gives-no-db/74829>

Thank you very much for the great series!

**Comment by mbrandonw** [comment]
Hi @morstin, I don't think there are any universal, correct solutions to these problems.

> How to conform the table structs as Player, Team to Identifiable?

While I understand why it can be strange to have an optional ID in an `Identifiable` conformance, I do think it is the more pragmatic approach to just do it anyway. SwiftUI demands identifiable types, and auto-increment keys demands optionality. So you are stuck between a rock and a hard place. 

So I personally think it is ok to conform to `Identifiable` and just remember that all un-persisted records have the same identity.

If you are willing to use macros you might be able to create some kind of `@Draftable` macro for creating "draft" versions of your model:

```swift
@Draftable
struct Player {
  var id: Int64
  var name: String
}

// expands to:
struct Player {
  var id: Int64
  var name: String

  struct Draft {
    var id: Int64?
    var name: String
  }
}
```

Then when creating players you can use `Player.Draft`. I still think there are substantial ergonomics problems to think through before this can be viable.

> How to keep database consistency when deleting?

I don't think we are going to have much to say about this. Only thing that comes to mind is adding a SQLite trigger to execute when songs are deleted to then clean up empty albums. But this really isn't going to be the focus of our explorations.

**Reply by morstin** [reply]
@mbrandonw Thank you for your quick response; I really appreciate it!
